ui/utility.py

```diff
@@ -9,17 +9,43 @@
 import re
 import logging
 import functools
+import tarfile
+import shutil
+from datetime import datetime
 log = logging.getLogger(__name__)
 
 class FormfeedConversionException(Exception): pass
 class LinefeedConversionException(Exception): pass
 
-def find_firmware(directory):
+def update_ui(config):
+    log.info("updating UI")
+    usb_dir = config.get('files', 'usb_dir')
+    ui_file = find_ui_update(usb_dir)
+
+    log.info("looking for new ui in %s" % usb_dir)
+    install_dir = config.get('files', 'install_dir')
+    install_dir = os.path.expanduser(install_dir)
+
+    archive_dir = config.get('files', 'archive_dir')
+    archive_dir = os.path.expanduser(archive_dir)
+
+    # archive old ui
+    archive_dir += datetime.now().strftime("%Y%m%d-%H%M%S-ui")
+    log.info("archiving %s to %s" % (install_dir, archive_dir))
+    shutil.move(install_dir, archive_dir)
```

Comment:
I don't see how it's not removing the files of a running program, even if they are being moved together. The running process thinks it's then running in a non-existent directory. I believe you that it works of course, but it seems precarious. 

The whole thing is a bit challenging even with a separate updater application because what if we want to update the updater application? The updater is analogous to a firmware bootloader, we can't really field upgrade it without things getting weird. 

We wouldn't need an extra file though,  we can read the state this way:
```
$ pwd
/home/kaspar/projects/braille/code/canute-ui
$ python
>>> from ui import initial_state
>>> initial_state.read()
<frozendict {'warming_up': False, 'replacing_library': False, 'menu': <frozendict {'data': ([23, 17, 15, 7, 1, 9, 17, 0, 7, 10, 3, 23, 1, 23, 61, 0, 11, 23, 21, 13, 0, 37, 14, 3, 0, 14, 30, 10, 9, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [14, 19, 37, 30, 25, 21, 58, 29, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [3, 1, 9, 5, 37, 15, 0, 7, 21, 27, 0, 30, 21, 0, 37, 14, 3, 0, 14, 30, 10, 9, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [23, 37, 29, 0, 58, 1, 23, 13, 0, 37, 15, 0, 23, 21, 37, 30, 10, 29, 17, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [23, 17, 14, 17, 30, 0, 25, 10, 14, 15, 7, 1, 61, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]), 'page': 0}>, 'resetting_display': False, 'library': <frozendict {'data': (), 'page': 0}>, 'books': (), 'backing_up_log': False, 'location': 'library', 'shutting_down': False, 'display': <frozendict {'width': 40, 'height': 9}>}>
>>> 
```

(We can't yet do `import ui; ui.initial_state.read()` for some reason though. This is kind of what I was rambling on about on Skype, `ui` is already a module, just not very well defined or exported. )

We can think about how to present the failed state to the user later, if you want to add some info I think it's a good idea but we likely won't make use of it in user-facing code for quite a while. For me it was just important to get into the mental model that this is a `failed` state and what to do about it.